In [1]:
import numpy as np

In [2]:
n = 30

In [3]:
inputs = [
    6,
    3.3333,
    0.3333,
    0.15,
    1.2609,
    0.875,
    0.5789,
    2,
    2.1429,
    0.8182,
    1.1818,
    3.75,
    0.5556,
    0.6923,
    2.5,
    2.1667,
    1.4286,
    0.5,
    8,
    2.3333,
    1.6667,
    0.4,
    2.6667,
    1.6667,
    1.2,
    0.5,
    2,
    0.8571,
    0.9333,
    6
]

In [4]:
x_mean = sum(inputs) / n
x_mean

1.94971

In [5]:
dispersion = sum((x - x_mean) ** 2 for x in inputs) / n
dispersion

3.3401709842333336

Общая дисперсия $D = 3.34$ характеризует меру рассеивания $x_i$ относительно общей средней $\overline{x} = 1.95$. 
Чем дисперсия больше, тем дальше разбросаны $x_i$ от средней $\overline{x}$

In [6]:
k = int(1 + 3.322 * np.log10(n))
k

5

In [7]:
h = (max(inputs) - min(inputs)) / k
h

1.5699999999999998

In [8]:
inputs = sorted(inputs)
inputs

[0.15,
 0.3333,
 0.4,
 0.5,
 0.5,
 0.5556,
 0.5789,
 0.6923,
 0.8182,
 0.8571,
 0.875,
 0.9333,
 1.1818,
 1.2,
 1.2609,
 1.4286,
 1.6667,
 1.6667,
 2,
 2,
 2.1429,
 2.1667,
 2.3333,
 2.5,
 2.6667,
 3.3333,
 3.75,
 6,
 6,
 8]

In [9]:
groups = [inputs[i: i + k + 1] for i in range(0, len(inputs), k + 1)]
groups

[[0.15, 0.3333, 0.4, 0.5, 0.5, 0.5556],
 [0.5789, 0.6923, 0.8182, 0.8571, 0.875, 0.9333],
 [1.1818, 1.2, 1.2609, 1.4286, 1.6667, 1.6667],
 [2, 2, 2.1429, 2.1667, 2.3333, 2.5],
 [2.6667, 3.3333, 3.75, 6, 6, 8]]

In [10]:
groups[0].append(groups[1].pop(0))
groups[-2].append(groups[-1].pop(0))
groups

[[0.15, 0.3333, 0.4, 0.5, 0.5, 0.5556, 0.5789],
 [0.6923, 0.8182, 0.8571, 0.875, 0.9333],
 [1.1818, 1.2, 1.2609, 1.4286, 1.6667, 1.6667],
 [2, 2, 2.1429, 2.1667, 2.3333, 2.5, 2.6667],
 [3.3333, 3.75, 6, 6, 8]]

In [11]:
n_i = [len(group) for group in groups]
n_i

[7, 5, 6, 7, 5]

In [12]:
x_mean_i = [sum(group) / len(group) for group in groups]
x_mean_i

[0.43111428571428567, 0.83518, 1.4007833333333333, 2.2585142857142855, 5.41666]

In [13]:
squared_deviation = [sum((x - x_mean_i[i]) ** 2 for x in group) for i, group in enumerate(groups)]
squared_deviation

[0.13638876857142854,
 0.032396667999999997,
 0.25003210833333345,
 0.3859796285714286,
 14.472361112]

In [14]:
dispersion_i = [squared_dev / n_i[i] for i, squared_dev in enumerate(squared_deviation)]
dispersion_i

[0.019484109795918362,
 0.0064793336,
 0.04167201805555557,
 0.05513994693877551,
 2.8944722224]

Групповая дисперсия характеризует меру разброса значений группы, относительно групповой средней. Наименьшая дисперсия получилась по 2-й группе, это означает, что варианты $x_i$ этой группы расположены достаточно близко к средне групповой. Максимальная дисперсия в 5 группе, это означает, что варианты $x_i$ этой группы расположены далеко от средней групповой.

In [ ]:
ingroup_dispersion = sum((dispersion_i[i]) * n_i[i] for i in range(len(n_i))) / n
ingroup_dispersion

2.830932374717461

Внутригрупповая дисперсия характеризует средневзвшанную вариацию значений $x_i$ по группам.

In [22]:
a = [(x_mean - x_mean_i[i]) ** 2 for i in range(len(x_mean_i))]
a

[2.3061329434469395,
 1.2421771209,
 0.3013204853777779,
 0.09536008687551002,
 12.019742302500001]

In [23]:
squared_deviation_n_i = [squared_dev * n_i[i] for i, squared_dev in enumerate(a)]
squared_deviation_n_i

[16.142930604128576,
 6.2108856045000005,
 1.8079229122666673,
 0.6675206081285702,
 60.0987115125]

In [24]:
aroundgroup_dispersion = sum(squared_deviation_n_i) / n
aroundgroup_dispersion

2.830932374717461

Межгрупповая дисперсия характеризует меру разброса групповых средних относительно общих средних. Чем эта дисперсия больше, тем дальше расположены групповые средние от общей средней.

Этот коэффициент харктеризует долю вариации, объяснённую группировочным фактором. Коэффициент детерминации

Чем больше межгрупповая дисперсия, тем больше коэффициент детерминации

85 % металлоёмкости станков объясняются тем фактором, что станки принадлежат разным производителям, оставшиеся 15% вариации объясняются другими причинами